BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m

```diff
@@ -0,0 +1,297 @@
+            BOXAssert(error != nil, @"Failed to create cacheDir %@ with error %@", cacheDir, error);
+        }
+        self.cacheDir = cacheDir;
+    }
+    return self;
+}
+
+- (BOOL)cacheUserId:(NSString *)userId associateId:(NSString *)associateId backgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId error:(NSError **)outError
+{
+    //persist users/$userId/$associateId/$backgroundSessionId-$sessionTaskId
+    BOOL success = [self createFileForUserId:userId associateId:associateId backgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId error:outError];
+
+    if (success == YES) {
+        //persist sessions/$backgroundSessionId/$sessionTaskId
+        success = [self createDirForBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId error:outError];
+    }
+
+    return success;
+}
+
+- (BOOL)cacheBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId destinationFilePath:(NSString *)destinationFilePath error:(NSError **)error
+{
+    NSData *data = [destinationFilePath dataUsingEncoding:NSUTF8StringEncoding];
+    return [self cacheBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId data:data type:BOXURLSessionTaskCacheFileTypeDestinationFilePath error:error];
+}
+
+- (BOOL)cacheBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId responseData:(NSData *)responseData error:(NSError **)error
+{
+    return [self cacheBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId data:responseData type:BOXURLSessionTaskCacheFileTypeResponseData error:error];
+}
+
+- (BOOL)cacheBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId resumeData:(NSData *)resumeData error:(NSError **)error
+{
+    return [self cacheBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId data:resumeData type:BOXURLSessionTaskCacheFileTypeResumeData error:error];
+}
+
+- (BOOL)cacheBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId response:(NSURLResponse *)response error:(NSError **)error
+{
+    NSData *data = [NSKeyedArchiver archivedDataWithRootObject:response];
+    return [self cacheBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId data:data type:BOXURLSessionTaskCacheFileTypeResponse error:error];
+}
+
+- (BOOL)cacheBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId taskError:(NSData *)taskError error:(NSError **)error
+{
+    NSData *data = [NSKeyedArchiver archivedDataWithRootObject:taskError];
+    return [self cacheBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId data:data type:BOXURLSessionTaskCacheFileTypeError error:error];
+}
+
+- (BOXURLSessionTaskCachedInfo *)getCacheForUserId:(NSString *)userId associateId:(NSString *)associateId error:(NSError **)error
```

Comment:
+1 we should not be using 'get' in Obj-C generally speaking, it has an association with inserting the value into a variable passed in by reference. If you want to indicate that it's being loaded from disk, etc, you can use 'fetch...'